Fix singleton issue when leaving several nodes, #27487

patriknw opened the pull request:
* When leaving several nodes at about the same time the new singleton
  could be started before previous had been completely stopped.
* Found two possible ways this could happen.
  * Acting on MemberRemoved that is emitted when the self
    cluster node is shutting down.
  * The HandOverDone confirmation when in Younger state,
    but that node is also Leaving so could be seen as Exiting
    from a third node that is the next singleton.

Refs #27487

patriknw committed:
Fix singleton issue when leaving several nodes, #27487

* When leaving several nodes at about the same time the new singleton
  could be started before previous had been completely stopped.
* Found two possible ways this could happen.
  * Acting on MemberRemoved that is emitted when the self
    cluster node is shutting down.
  * The HandOverDone confirmation when in Younger state,
    but that node is also Leaving so could be seen as Exiting
    from a third node that is the next singleton.

johanandren commented on akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala:
```diff
@@ -648,9 +648,15 @@ class ClusterSingletonManager(singletonProps: Props, terminationMessage: Any, se
-      sender() ! HandOverDone
+      val selfStatus = cluster.selfMember.status
+      if (selfStatus == MemberStatus.Leaving || selfStatus == MemberStatus.Exiting)
+        logInfo("Ignoring HandOverToMe in Younger from [{}] because self is [{}].", sender().path.address, selfStatus)
```
Hmm, so we are on the previous oldest, because we are leaving, and got a request to hand over from another node, won't this always be true? 

Isn't the problem about the status of the node requesting the handover when `HandOverDone` arrives?

johanandren replied:
Sorry, read the explanation above now. 👍

patriknw replied:
No, the purpose of this is as commented a few lines down:
```
        // this node was probably quickly restarted with same hostname:port,
        // confirm that the old singleton instance has been stopped
```

This is in state Younger.

The way it was triggered for the failure scenario was that node-1 was the singleton, and then leaving node-1 and node-2. Then node3 could see node-2 as exiting and therefore believe that node-2 was previous oldest instead of node-3. But, node-2 itself had not seen the exiting yet so it was still in Younger and confirmed the request from node-3 although the real singleton actor was still running on node-1.

helena replied:
I'm wondering if this is the issue I'd seen before, where there were > 1 instances. Would be fantastic if it is, and this prevents that.

johanandren commented on akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala:
```diff
@@ -686,7 +692,9 @@ class ClusterSingletonManager(singletonProps: Props, terminationMessage: Any, se
-        if m.uniqueAddress == previousOldest =>
+        if m.uniqueAddress == previousOldest && !cluster.isTerminated =>
+      // don't act on DelayedMemberRemoved (starting singleton) if this node is shutting its self down,
+      // just wait for self MemberRemoved
```
:+1:

raboof commented on akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala:
```diff
@@ -686,7 +692,9 @@ class ClusterSingletonManager(singletonProps: Props, terminationMessage: Any, se
     case Event(DelayedMemberRemoved(m), BecomingOldestData(Some(previousOldest)))
-        if m.uniqueAddress == previousOldest =>
+        if m.uniqueAddress == previousOldest && !cluster.isTerminated =>
+      // don't act on DelayedMemberRemoved (starting singleton) if this node is shutting its self down,
```
I wonder if we should start ignoring those even earlier, already when `Leaving`?

patriknw replied:
I didn't want to change more than necessary. It may stay in Leaving (not moved to Exiting) if there are other unreachable and then it might be better to let it become singleton instead of not having an active singleton at all.

patriknw committed:
keep track of all previous oldest, not only the latest

* Option => List
* Otherwise in BecomingOldest it could transition to Oldest
  when the previous oldest was removed even though the previous-previous wasn't removed yet

patriknw commented:
When running repeat job the test revealed another scenario:

* N1: singleton started
* N1: leaving
* N2: leaving
* N1: observed oldest N1 -> N3 => WasOldest
* N2: observed oldest N1 -> N2 => BecomingOldest
* N3: observed oldest N1 -> N2
* N3: observed oldest N2 -> N3 => BecomingOldest
* N2: Exiting, Exiting completed
* N3: Previous oldest [N2] removed => Oldest

The last step at N3 means that the singleton is started there even though the singleton has still not stopped completely at N1.

Since it was only keeping track of one previous oldest in an `Option` it took the removal of N2 as confirmation that previous singleton was stopped, although it was still running at N1.

The solution is to keep track of all previous older nodes in a `List` instead of the `Option`. All previous must be removed before transitioning from BecomingOldest to Oldest in above scenario.  Fixed in 2db87d5

patriknw committed:
fix failure in ClusterSingletonRestart2Spec

* OldestChanged was not emitted when Exiting member was removed
* The initial membersByAge must also contain Leaving, Exiting members

patriknw committed:
mima filter

johanandren commented on akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala:
```diff
@@ -597,36 +601,42 @@ class ClusterSingletonManager(singletonProps: Props, terminationMessage: Any, se
+        if (previousOldest.isEmpty)
+          tryGotoOldest()
+        else if (previousOldest.forall(removed.contains))
+          tryGotoOldest()
```
Perhaps fold this into a single `if (previousOldest.isEmpty || previousOldest.forall(removed.contains)) tryGotoOldest else handover` - could also be expressed as a single `previousOldest.forall(removed.contains)`

patriknw replied:
fixed, changed to single forall

johanandren reviewed (approved):
LGTM with one nit

patriknw committed:
review feedback

helena commented on akka-cluster-tools/src/multi-jvm/scala/akka/cluster/singleton/ClusterSingletonManagerLeave2Spec.scala:
```diff
@@ -0,0 +1,200 @@
+        expectMsg("postStop")
+        // CoordinatedShutdown makes sure that singleton actors are
+        // stopped before Cluster shutdown
+        expectMsg(10.seconds, "MemberRemoved")
```
This is the test I was hoping to see, the MemberRemoved check.

helena reviewed (approved):
LGTM

patriknw commented:
I'll forward port this to master
